### ISSUE TITLE
ARM build: No warning raised matching: divide by zero, invalid value

### ISSUE THREAD

**Issue by olebole** [scikit-image/scikit-image#3335]
During the Debian build, we observe the following [test error](https://buildd.debian.org/status/fetch.php?pkg=skimage&arch=armel&ver=0.14.0-1&stamp=1530063617&raw=0) on arm (32 bit, little endian, software floating point). The sister architecture with hardware floating point does not have this problem (but fails with #3092).
```
_______________________________ test_shape_index _______________________________

    def test_shape_index():
        square = np.zeros((5, 5))
        square[2, 2] = 4
        with expected_warnings(['divide by zero', 'invalid value']):
>           s = shape_index(square, sigma=0.1)

../debian/tmp/usr/lib/python2.7/dist-packages/skimage/feature/tests/test_corner.py:174: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/lib/python2.7/contextlib.py:24: in __exit__
    self.gen.next()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

matching = ['divide by zero', 'invalid value']

    @contextmanager
    def expected_warnings(matching):
        """[...]"""
        with all_warnings() as w:
            # enter context
            yield w
            # exited user context, check the recorded warnings
            remaining = [m for m in matching if '\A\Z' not in m.split('|')]
            for warn in w:
                found = False
                for match in matching:
                    if re.search(match, str(warn.message)) is not None:
                        found = True
                        if match in remaining:
                            remaining.remove(match)
                if not found:
                    raise ValueError('Unexpected warning: %s' % str(warn.message))
            if len(remaining) > 0:
                msg = 'No warning raised matching:\n%s' % '\n'.join(remaining)
>               raise ValueError(msg)
E               ValueError: No warning raised matching:
E               divide by zero
E               invalid value

../debian/tmp/usr/lib/python2.7/dist-packages/skimage/_shared/_warnings.py:124: ValueError
```
Are the warnings really required, or may the not appear as well?

Putting @yarikoptic and @tillea in Cc

**Comment by hmaarrfk** [comment]
@olebole I was definitely agreesive with catching all warnings :

  * there is a second warning that is issued too, can you see what happens if you replace 
```
        with expected_warnings(['divide by zero', 'invalid value']):
```

with
```
        with expected_warnings(['invalid value']):
```

**Comment by hmaarrfk** [comment]
TLDR: Probably the best way is to add `r'|\A\Z'` to the expected warnings.

I had a chance to look through the code and the reference

The issue comes from the use of `arctan`:
https://github.com/scikit-image/scikit-image/blob/d1f048e1b55f64865cd3eba4a2b08c1680426f6a/skimage/feature/corner.py#L421

Unfortunately, the way the `shape` is defined, it is difficult (for me) to write a close form formula that uses some combination of `arctan2`, `modulo` or whatever else that avoids the use of the division operator. The reason is that the shape function is defined on a closed interval (in math notation) $$s \in [-1, 1]$$ and I can't figure out a clever way to make the C code give something in closed interval that I can map to the correct answer. It originates from the fact that `-0` and `+0`are different for IEEE, as are `-inf` and `+inf`.

Finally, currently the computation of the angle is order "cost of 1 division" and "cost of arctan". I wouldn't want to blow that up significantly for an edge case that is correctly handled by `arctan`

For reference, here is something you can use to visually check for correctness if you think of something smart:

```python
import numpy as np
from numpy import pi

the_sum = np.array([-4, -4, 0, 0, 4, 4])
the_diff = np.zeros(the_sum.shape)
negative_zero = -np.zeros(1)
the_diff[::2] = negative_zero  # -0 and +0 are actually different!!!

the_diff_corrected = the_diff
the_diff_corrected[the_diff > 0] *= -1

print('arctan2  : ' + str(np.arctan2(the_sum, the_diff)/pi))
print('arctan   : ' + str(np.arctan(the_sum/the_diff)/pi))
print('expected : ' + '[ 0.5 -0.5  dc   dc  -0.5  0.5]')
print('zero fun : ' + str(the_sum/the_diff))
print('         : ' + str(1 * the_diff))
print('         : ' + str(np.sign(the_diff)))
print("dc is don't care or some 0 value")
```
```
arctan2  : [-0.5 -0.5  1.   0.   0.5  0.5]
arctan   : [ 0.5 -0.5  nan  nan -0.5  0.5]
expected : [ 0.5 -0.5  dc   dc  -0.5  0.5]
zero fun : [ inf -inf  nan  nan -inf  inf]
         : [-0.  0. -0.  0. -0.  0.]
         : [0. 0. 0. 0. 0. 0.]
dc is don't care or some 0 value
```

**Comment by hmaarrfk** [comment]
@olebole I really don't like adding the blanket `optional warning` flag. Do you know how we can detect arm floating point to disable that check just for the desired architecture?

Does arm (software floating point) not raise these exception? If so, why not?

**Comment by olebole** [comment]
@hmaarrfk probably, one could use `sys.implementation` for this:
```
>>> sys.implementation
namespace(_multiarch='arm-linux-gnueabi', cache_tag='cpython-35', hexversion=50660336, name='cpython', version=sys.version_info(major=3, minor=5, micro=3, releaselevel='final', serial=0))
```
The "hard float" variant has `_multiarch='arm-linux-gnueabihf'` here.

**Comment by hmaarrfk** [comment]
I never wrote that actual function, but I did add the hard check for the warnings.

Unfortunately, the way the function is defined mathematically, the output should only be on the right half plane (if you think of the angle as being part of a complex number). It also includes the extremities of pi/2 and -pi/2 and those extreme points are actually quite important to capture (according to the paper). I couldn't find a way to obtain the correct behaviour for those using `arctan2`

See the output of the little script above. You are free to try, but as you can see, there are sign errors at $+-\pi/2$ It is a fun puzzle. I couldn't solve it.
